web/components/admin/AdminGate/index.js

```diff
@@ -0,0 +1,63 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Redirect } from 'react-router-dom';
+import { firebaseConnect, firestoreConnect, isLoaded, isEmpty } from 'react-redux-firebase';
+import { connect } from 'react-redux';
+import { compose } from 'redux';
+
+import { Logout } from '../../auth';
+
+const AdminGate = ({ auth, firestore, children }) => {
+  const {
+    status: {
+      requesting: {
+        admins: isAdminsLoading,
+      },
+    },
+  } = firestore;
+
+  if (!isLoaded(auth) || isAdminsLoading) {
+    return (<span>Loading...</span>);
+  }
+
+  const {
+    data: {
+      admins,
+    },
+  } = firestore;
+
+  const { uid } = auth;
+  const isAdmin = uid in admins;
+
+  if (isEmpty(auth)) {
+    return (<Redirect to="/page_not_found" />);
```

Comment:
Maybe we should redirect to `/admin/login` here instead of `page_not_found`

Reply:
hm maybe, not convinced why /admin/login is strictly better than page_not_found

Reply:
This condition is triggered when the user is trying to access an admin page, we can assume that it's an organizer who's trying to access a page but has forgotten to log in. In this case it would save them the trouble of navigating to `/admin/login` manually.

Reply:
Ex: navigating directly to `nwhacks.io/#/admin/assessment` through the dropdown history on a URL bar

Reply:
Yeah I agree, but it also makes the `admin` pages more likely to be noticed by non admin users. The behaviour of non admin users are unpredictable, but the behaviour of < 30 admin users is (i.e. we can explicitly say they need to login first at `/admin/login` before having access to the admin features)

Reply:
True but I don't see an issue with them discovering `/admin/` especially since if they were really trying to snoop around they would probably be looking at this repo for the routes. Redirecting to login seems like it would improve the UX experience for organizers instead of forcing them to use `/admin/login` each time or making them double check that they have no typos in their URL.

Reply:
Well, it's not just about malicious users, it's also about the experience of the non admin users who accidentally stumble upon the admin section and they are prompted to login, at which point they would be very confused and uncomfortable.

I think we can both agree that this is just a UX decision, and I don't have enough expertise to make the call on which user experience is more valuable (organizers vs non-admin users). What do you think about consulting the design team?

Reply:
Yeah that sounds like a good idea

Reply:
I'm a little late but some thoughts: If a normal user "accidentally" finds the login page then its either a crazy coincidence or we don't have url routes that make sense. I think its better UX to expect someone visiting a url to want to be there (which hangs on the fact that we have good routes/good UX in other parts of the website). A simple "back to nwhacks" would also help reduce confusion. I also don't think this is worth changing unless it causes problems in the future though.. just wanted to add my 2 cents.